rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,58 @@
+# RFC 192 - `minCapacity` ensures workers do not get unnecessarily killed
+* Comments: [#192](https://github.com/taskcluster/taskcluster-rfcs/pull/192)
+* Proposed by: @JohanLorenzo
+
+# Summary
+
+`worker-manager` allows `minCapacity` to be set, ensuring a certain number of workers
+are available at any given time. Unlike what current happens now, these workers
+shouldn't be killed unless `minCapacity` is exceeded.
+
+## Motivation - why now?
+
+As far as I can remember, the current behavior has always existed. This year, the
+Engineering Effectiveness org is optimizing the cost of the Firefox CI instance.
+[Bug 1899511](https://bugzilla.mozilla.org/show_bug.cgi?id=1899511) made a change that
+actually uncovered the problem with the current behavior: workers gets killed after 2
+minuted and a new one gets spawned.
+
+
+# Details
+
+In the current implementation, workers are in charge of knowning when they have to shut
+down. Given the fact `docker-worker` is officially not supported anymore and we can't
+cut a new release and use it, let's change what config `worker-manager` gives to all
+workers, `docker-worker` included.
+
+## When `minCapacity` is exceeded
+
+In this case, nothing should change. `worker-manager` sends the same config to workers
+as it always did.
+
+## When `minCapacity` is not yet met
+
+Here, `worker-manager` should increase `afterIdleSeconds` to a much higher value (e.g.:
+24 hours). This way, workers remain online long enough and we don't kill them too often.
+In case one of these long-lived workers get killed by an external factor (say: the
+cloud provider reclaims the spot instance), then `minCapacity` won't be met an a new
+long-lived one will be created.
+
+### What if we deploy new worker images?
+
+Long-lived workers will have to be killed if there's a change in their config, including
```

Comment:
This goes along with the #191 proposal quite well. 
With introduction of launch configurations, workers would be able to query w-m periodically to see if config is still active, and restart/shutdown if not. 

We could probably create a more generic endpoint which would answer if worker should terminate:
- `yes` if config changed
- `no` if `minCapacity` is not observed.

However, I'm not totally sure how to solve this issue if all workers at once ask if they could shutdown, and all get `yes`.. just to have zero running workers after

Reply:
> This goes along with the https://github.com/taskcluster/taskcluster-rfcs/pull/191 proposal quite well.

Nice! 👍 

> However, I'm not totally sure how to solve this issue if all workers at once ask if they could shutdown, and all get yes.. just to have zero running workers after

I guess that's an okay behavior, at least for a first iteration. Maybe `worker-manager` could remember how many workers are being shut down so it doesn't tell more workers to turn off.